### PR TITLE
Fix ConcurrentModificationException in listeners

### LINKED ISSUE #1970: ConcurrentModificationException while making notification
Hi,
i'm geting exception when running below code:
<pre><code>public static void makeMutableWorkoutCopy(Realm realm, final long workoutId) {
        realm.beginTransaction();
            Workout workout = realm.where(Workout.class).equalTo("id", workoutId).findFirst();

            Workout copyWorkout = realm.copyFromRealm(workout);
            copyWorkout.setId(IdentifierGenerator.getInstance().generateUniqueId(realm, Workout.class));
            copyWorkout.setName(copyWorkout.getName() + "*");
            copyWorkout.setSections(new RealmList<Section>());
            copyWorkout.setImmutableWorkout(workout);

            workout.setMutableWorkout(realm.copyToRealmOrUpdate(copyWorkout));
        realm.commitTransaction();
    }</code></pre>
Exception:
<pre><code>java.util.ConcurrentModificationException
at java.util.IdentityHashMap$IdentityHashMapIterator.checkConcurrentMod(IdentityHashMap.java:164)
at java.util.IdentityHashMap$IdentityHashMapIterator.next(IdentityHashMap.java:169)
at io.realm.HandlerController.notifyRealmObjectCallbacks(HandlerController.java:254)
at io.realm.HandlerController.notifyTypeBasedListeners(HandlerController.java:207)
at io.realm.BaseRealm.commitTransaction(BaseRealm.java:294)
at io.realm.Realm.commitTransaction(Realm.java:108)
at com.borg.forza.model.RealmUtils.makeMutableWorkoutCopy(RealmUtils.java:289)
</code></pre>

The model of Workout.class looks like below:
<pre><code>public class Workout extends RealmObject {

    @Index      private long                    id;
                private long                    backendId;
                private int                     sync;

    @PrimaryKey private String                  name;
                private String                  description;
    @Index      private String                  type;
                private long                    duration;
                private RealmList<Section>      sections;
                private int                     access;
                private Workout                 mutableWorkout;
                private Workout                 immutableWorkout;
    }</code></pre>
I am using 0.87.0-SNAPSHOT.
Thanks!

### PR DESCRIPTION
**Pull request by beeender**
* It should be legal to modify realmObjects map in listener.
 * It should be legal to modify syncRealmResults/asyncRealmResults map
   in listener.

Fix #1970

### CODE REVIEW

**Comment by beeender** [comment]
@realm/java

**Comment by beeender** [comment]
retest this please

**Comment by beeender** [comment]
retest this please

**Comment by dalinaum** [comment]
retest this please

**Comment by dalinaum** [comment]
retest this please

**Comment by kneth** [comment]
@realm/java Please review

File: changelog.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 0.87.2
  * Fixed a bug when RealmObjectSchema.addField called with PRIMARY_KEY modifier, the field was not set as a required field (#2001).
+ * Fixed a bug which could cause ConcurrentModificationException when get a new RealmObject in RealmObject's change listener (#1970).
```
Comment (cmelchior) [inline changelog.txt]: Not entirely sure what should be here, but "get" is not right

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 2]
```diff
@@ -1807,4 +1809,69 @@ public void run() {
...
+                realm.beginTransaction();
+                final Owner owner = realm.createObject(Owner.class);
+                owner.setCat(realm.createObject(Cat.class));
+                realm.commitTransaction();
+
+                owner.addChangeListener(new RealmChangeListener() {
+                    @Override
+                    public void onChange() {
+                        Cat cat = owner.getCat();
+                        // Keep calling addToRealmObjects just in case it won't be called in getCat in the future
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: This seems a bit dangerous. Isn't it better that this unit test start failing if addToRealmObjects no longer gets called?

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 3]
```diff
@@ -1807,4 +1809,69 @@ public void run() {
...
+                        realm.handlerController.addToRealmObjects(cat);
+                        signalTestFinished.countDown();
+                    }
+                });
+
+                handler.post(new Runnable() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: Why posting again?

File: realm/realm-library/src/main/java/io/realm/HandlerController.java [hunk 4]
```diff
@@ -236,19 +236,26 @@ private void notifySyncRealmResultsCallbacks() {
     }
 
     private void notifyRealmResultsCallbacks(Iterator<WeakReference<RealmResults<? extends RealmObject>>> iterator) {
+        List<RealmResults<? extends RealmObject>> resultsToBeNotified =
+                new ArrayList<RealmResults<? extends RealmObject>>();
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: Can't this fit on one line?

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 5]
```diff
@@ -1807,4 +1809,69 @@ public void run() {
...
+        handler.post(new Runnable() {
+            @Override
+            public void run() {
+                realm = Realm.getInstance(configuration);
+
+                RealmResults<Owner> results = realm.allObjects(Owner.class);
+                results.addChangeListener(new RealmChangeListener() {
+                    @Override
+                    public void onChange() {
+                        realm.handlerController.addToRealmResults(realm.allObjects(Owner.class));
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: Why is this needed?

File: changelog.txt (hunk 1 above)
Comment (beeender) [inline changelog.txt]: @cm what about:
Fixed a bug which could cause ConcurrentModificationException when manipulate with RealmObject in RealmObject's change listener (#1970).

File: realm/realm-library/src/main/java/io/realm/HandlerController.java (hunk 4 above)
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: Nope, i have a border line set in my IDE. So normally i am right :grin:

Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: :+1:

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 6]
```diff
@@ -1807,4 +1811,89 @@ public void run() {
...
+    // Test modifying realmObjects in RealmObject's change listener
+    public void test_change_realm_objects_map_in_listener() throws InterruptedException {
+        final CountDownLatch finishedLatch = new CountDownLatch(2);
+        handler.post(new Runnable() {
+            @Override
+            public void run() {
+                realm = Realm.getInstance(configuration);
+
+                realm.beginTransaction();
+                // At least two objects are need to make sure list modification happen during iterating.
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: > need

needed

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 7]
```diff
@@ -1807,4 +1811,89 @@ public void run() {
...
+                RealmChangeListener listener = new RealmChangeListener() {
+                    @Override
+                    public void onChange() {
+                        Cat cat = owner.getCat();
+                        boolean foundKey = false;
+                        // Keep calling addToRealmObjects just in case it won't be called in getCat in the future
+                        for (WeakReference<RealmObject> weakReference : realm.handlerController.realmObjects.keySet()) {
+                            if (weakReference.get() == cat) {
+                               foundKey = true;
+                                break;
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: remove space before `break`

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 8]
```diff
@@ -1807,4 +1811,89 @@ public void run() {
...
+                final Owner owner = realm.createObject(Owner.class);
+                owner.setCat(cat);
+                realm.commitTransaction();
+
+                RealmChangeListener listener = new RealmChangeListener() {
+                    @Override
+                    public void onChange() {
+                        Cat cat = owner.getCat();
+                        boolean foundKey = false;
+                        // Keep calling addToRealmObjects just in case it won't be called in getCat in the future
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: comment is ambiguous + the test comment state that we're modifying the RealmObject inside the listener, which we're not doing obviously?

**Comment by nhachicha** [comment]
besides minor typo :+1:

**Comment by dalinaum** [comment]
:+1:

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 9]
```diff
@@ -1807,4 +1811,89 @@ public void run() {
...
+                final Cat cat = realm.createObject(Cat.class);
+                final Owner owner = realm.createObject(Owner.class);
+                owner.setCat(cat);
+                realm.commitTransaction();
+
+                RealmChangeListener listener = new RealmChangeListener() {
+                    @Override
+                    public void onChange() {
+                        Cat cat = owner.getCat();
+                        boolean foundKey = false;
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: Why do this? It sounds like testing an implementation detail? I would assume that:

```
Cat cat  = owner.getCat()
finishedLatch.countDown()
```

Would be enough?

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java [hunk 10]
```diff
@@ -1807,4 +1811,89 @@ public void run() {
...
+                realm = Realm.getInstance(configuration);
+
+                // Two results needed to make sure list modification happen while iterating
+                RealmResults<Owner> results1 = realm.allObjects(Owner.class);
+                RealmResults<Cat> results2 = realm.allObjects(Cat.class);
+                RealmChangeListener listener = new RealmChangeListener() {
+                    @Override
+                    public void onChange() {
+                        RealmResults<Owner> results = realm.allObjects(Owner.class);
+                        boolean foundKey = false;
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: Same as above

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java (hunk 9 above)
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTest.java]: I would like to keep that in case of we change the `getCat()` behaviour in the future. eg.: only add it to the list when adding change listener.

**Comment by cmelchior** [comment]
:+1:
